stdlib/public/core/FloatingPointToString.swift

```diff
@@ -0,0 +1,1291 @@
+
+        let one = UInt64(1) << (64 - adjustIntegerBits)
+        let adjustFractionMask = one - 1;
+        let oneHalf = one >> 1;
+        if deltaHigh64 >= tHigh64 &+ one {
+            // The `skew` is the difference between our
+            // computed digits and the original exact value.
+            var skew: UInt64
+            if (d.significandBitPattern == 0) {
+                skew = deltaHigh64 &- deltaHigh64 / 3 &- tHigh64
+            } else {
+                skew = deltaHigh64 / 2 &- tHigh64
+            }
+
+            // We use the `skew` to figure out whether there's
+            // a better base-10 value than our current one.
+            if (skew & adjustFractionMask) == oneHalf {
+                // Difference is an integer + exactly 1/2, so ...
+                let adjust = skew >> (64 - adjustIntegerBits)
+                var t = unsafe buffer.unsafeLoad(fromUncheckedByteOffset: nextDigit - 1,
+                                                 as: UInt8.self)
+                t &-= UInt8(truncatingIfNeeded: adjust)
+                // ... we round the last digit even.
+                t &= ~1
+                unsafe buffer.storeBytes(of: t,
+                                         toUncheckedByteOffset: nextDigit - 1,
+                                         as: UInt8.self)
+                // Note: "exactly" 1/2 is a subtle point above; this
+                // determination relies on various roundings canceling
+                // out, and proving correctness requires proper
+                // testing.  Testing so far has validated the
+                // correctness of this code.  However, even if that
+                // were not true, this only affects whether we choose
+                // the theoretically-ideal even final digit when an
+                // odd final digit would otherwise satisfy all
+                // requirements.
+            } else {
+                let adjust = (skew + oneHalf) >> (64 - adjustIntegerBits)
+                var t = unsafe buffer.unsafeLoad(fromUncheckedByteOffset: nextDigit - 1,
+                                                 as: UInt8.self)
+                t &-= UInt8(truncatingIfNeeded: adjust)
+                unsafe buffer.storeBytes(of: t,
+                                         toUncheckedByteOffset: nextDigit - 1,
+                                         as: UInt8.self)
+            }
+        }
+    }
+
+    // Step 8: Finalize formatting by rearranging
+    // the digits and filling in decimal points,
+    // exponents, and zero padding.
+    let isBoundary = (d.significandBitPattern == 0)
+    let forceExponential = (binaryExponent > 54) || (binaryExponent == 54 && !isBoundary)
+    return finishFormatting(&buffer, d.sign, firstDigit, nextDigit,
+                            forceExponential, base10Exponent)
+}
+
+@available(macOS 9999, *)
+// TODO: This doesn't guarantee inlining in all cases :(
+@inline(__always)
```

Comment:
@eeckstein Why might this inlining fail in release builds?   I've seen it both succeed and fail, possibly relating to whether I have enabled some local changes that force-inline a bunch of MutableSpan APIs that are used here.  Is it possible that this is close to some inlining limit?

Reply:
>  Is it possible that this is close to some inlining limit?

Perhaps related to the depth of nesting?

Reply:
> Why might this inlining fail in release builds? 
> Is it possible that this is close to some inlining limit?

yes, see https://github.com/swiftlang/swift/blob/439afd640693cf45f2d2f5f60f4c6586d7020ca9/lib/SILOptimizer/Transforms/PerformanceInliner.cpp#L932

Reply:
> Why do you need `@inline(__always)` here? Is it required to overrule SIL and LLVM inlining heuristics for performance?

Thank you for that pointer!   Why is the limit 64 there?  Was there some testing that went into that choice?

For the C version, inlining this function does seem to make a big difference:  This function has a lot of cases, but only one case is used at any one time, so the cost of setting up 7 function arguments in the caller does seem to be a significant portion of the runtime.  The biggest problem so far has been that the arbitrary limits on `@inline(__always)` make it hard to experiment.  I would like to at least try inlining to see what impact it has.

Reply:
The original motivation of `@inline(__always)` was to tweak the inlining decision of some small, time-critical, stdlib functions. It was not meant to be a user-visible attribute. Nevertheless, people started to (mis-)use it, e.g. by placing it everywhere they thought it might affect performance. But in reality, this does not work because it can easily blow up the compiler. Therefore we added this quite strict limits.
`@inline(__always)` is not what it sounds like because it's name is misleading.